### ISSUE TITLE
Accessibility issues related to PDF

### ISSUE THREAD

**Issue by Febakke** [Altinn/app-lib-dotnet#196]
### Description of the bug

Used Adobe Acrobat Reader to test our new pdfs and a couple of points are recurring.

**1. There is no defined primary language**
In the same way as you need to define `lang` in HTML, a PDF can also have its language defined. This is handy for screen readers. [PDF WCAG technique](https://www.w3.org/WAI/WCAG22/Techniques/pdf/PDF16)

**2. There is no title element**
If possible we should use the same title as the app does. [PDF WCAG technique](https://www.w3.org/TR/WCAG20-TECHS/PDF18.html)

**3. Tab order**
The accessibility test gives an error here. Im not sure how we can fix this because there are noe focusable elements in the pdf. Maybe just flag this as not relevant? [PDF WCAG Technique](https://www.w3.org/TR/WCAG20-TECHS/PDF3.html)

### Steps To Reproduce

Create a PDF
Use Adobe Acrobat Reader and check accessibility
See list of errors

**Comment by bjosttveit** [comment]
After some investigation into these issues, it looks to me as if browserless/puppeteer/chromium can't fix these for us. It can therefore not be done anything about in app-frontend either. The first two issues (I think) could be fixed on the backend, by doing some additional processing to the PDF-file that is returned from the PDF service before it is saved to storage. What needs to happen there is adding some metadata to the document by for example using a PDF library in .NET. The last issue appears to have something to do with chromium not setting the tags correctly. I tried generating a PDF from a very simple html-document and it still had the same issue, suggesting that no changes to app-frontend would make things better. The third issue I think would be difficult for us to fix, as it probably entails manually changing the structure of the PDF-tags, and I am not sure that this is easily achieved using a PDF-library or even possible at all.

I will transfer this issue to app-lib-dotnet as that is where the changes need to happen.

**Comment by bjosttveit** [comment]
Btw, [Pave](https://pave-pdf.org/pave/index.html) appears to be a reasonably good free tool for checking PDF accessibility.

**Comment by FinnurO** [comment]
Informasjon fra Chromium: https://blog.chromium.org/2020/07/using-chrome-to-generate-more.html

**Comment by FinnurO** [comment]
@bjosttveit Could you follow up #1 and #2 with Chromium bug system? https://bugs.chromium.org/p/chromium/issues/detail?id=1362536&q=wcag&can=2

And consider reporting #3 as bug if not already reported?

**Comment by bjosttveit** [comment]
Language and title issues should be fixed in Chrome version 123.0.6268.0 and later
https://bugs.chromium.org/p/chromium/issues/detail?id=1362536#c22

**Comment by bjosttveit** [comment]
I did some local testing of the version of the PDF-generator defined in the charts (with `puppeteer-21.4.1`). In my case that had a chromium version of `119.0.6045.9`. I tested in Pave, and it seems like it sets the tags much better now than when I tested this one year ago. The only issues now in my PDF are missing alternative texts in some images and links, which is probably easily fixed in my app configuration. (except for language and title which needs an even newer version)

I have not tested using a screen reader, but I can imagine it should be much better since the "reading order" listed in Pave looks really good now where it did not before. These issues should therefore already be fixed in prod 🥳 

As for language and title, I will test again when an up-to-date enough chromium version is shipped with browserless that I can test (there are some issues with apple-silicon on the latest images currently). We need to make sure the frontend will set the correct language and title when this becomes possible.

Additionally, it looks like chromium has added support for generating a document outline in version 121 I think. Which produces links in the sidebar (usually) of PDF readers that point to headings (see [this](https://github.com/puppeteer/puppeteer/pull/11779#issue-2106034157)). This may require setting some flags in app-lib to enable.

**Comment by bjosttveit** [comment]
Chromium version 123 is released and stable. It is available in puppeteer v22.6+. However, the browserless images still seem to be several versions behind on dockerhub, and its not clear if or when these will be updated. We are currently using v1 of browserless, and since then, they have released v2 and moved to GitHub Container Registry. v2 seems to be more up to date with puppeteer.

To get more up-to-date chromium/puppeteer versions, we may need to consider either migrating to browserless v2, or making our own PDF-service using puppeteer directly which would give us more control over the PDF-generation process anyway.

**Comment by bjosttveit** [comment]
Status update:
Browserless v1 has still not gotten any updated puppeteer/chromium version, so we are still on puppeteer version 21.4.1 and chromium version 118/119 something. At this point we should not expect v1 to get any further updates. 

The tab-order issue looks like it was mostly solved in this version, but note that in local-test we are on version 21.3.6 instead because there was no arm64 builds for the latest one, so generating PDF in localtest has some more issues and should not be used for testing this. As I mentioned previously there are still issues related to links and images, but I don't know if this is due to chromium or if we just don't have accessible links 🤷 

The missing title and language issues are of course still present as these were not fixed in chromium until version 123. (https://issues.chromium.org/issues/40864319#comment23)

As mentioned in my previous comment, since we cannot expect patches of browserless v1 we need to either update to v2, or create our own. I took a look at the v2 API and compared it to v1:

- [v1](https://chrome.browserless.io/docs/#/Browser%20API/post_pdf)
- [v2](https://docs.browserless.io/open-api#tag/Browser-REST-APIs/paths/~1chrome~1pdf/post)

Unfortunately, there are breaking changes (like `waitFor`) between these that affects us and would require us to update app-lib-dotnet to make it work. We therefore cannot simply drop in v2 as a replacement as it would require app owners upgrading their apps after we make the necessary changes.

So making our own may make migration easier, as we could keep our API backwards-compatible with browserless v1 if we wanted, so that we could just swap out browserless v1 without first upgrading all apps to be compatible.

**Comment by Magnusrm** [comment]
Based on the analysis done for this issue we need a discussion that results in a decision about how we should implement a solution.

**Comment by bjosttveit** [comment]
Since pdf3 was launched we are pretty much up-to-date on the latest chrome versions again, and tagged pdf generation was enabled in https://github.com/Altinn/altinn-studio/pull/17054.

I suspect that accessibility checkers will still complain about tag nesting. But there is really nothing we can do about it at this point (besides manual post-processing the PDFs). However, I don't think this will be an issue in practice for screen readers. Language and title issues should be resolved now, and the tags should be much better than when this issue was created.

One final thing we should consider doing, is changing the `<title>` element (which controls the document title) when we generate PDFs. This does not get set to anything particular by app-frontend (it does normally, just not in PDF mode), and simply becomes `{org}- {app}` which is from [Index.cshtml](https://github.com/Altinn/altinn-studio/blob/a470b27fedb54d6d9a3d7397eb03b203cb4cddc6/src/App/template/src/App/views/Home/Index.cshtml#L9).

Another bonus thing we can consider, is to generate a document outline that gets used by PDF-readers like this:

<img width="303" height="278" alt="Image" src="https://github.com/user-attachments/assets/4fba5d18-47cd-4ada-b892-04ae20f56ada" />

To get this we simply need to set `"generateDocumentOutline": true` in the CDP command to chrome. This gets generated automatically based on headings.
